File: 2023/Day 1/php/day1.php
<?php
// This is built on top of Rawoke's attempt. I simply optimised it slightly.

$r1 = $r2 = 0;

$lines = file('input.txt');

function get_num($line)
{
    static $num_lookup = [
        "zero" => "0",
        "one" => "1",
        "two" => "2",
        "three" => "3",
        "four" => "4",
        "five" => "5",
        "six" => "6",
        "seven" => "7",
        "eight" => "8",
        "nine" => "9",
    ];

    $pattern = '/(' . implode("|", array_keys($num_lookup)) . '|' . implode("|", array_values($num_lookup)) . ')/';
    $matches = [];

    if (preg_match_all($pattern, $line, $matches, PREG_OFFSET_CAPTURE)) {
        $firstNum = null;
        $lastNum = null;

        foreach ($matches[0] as $match) {
            $num = isset($num_lookup[$match[0]]) ? $num_lookup[$match[0]] : $match[0];

            if ($firstNum === null) {
                $firstNum = $num;
            }
            $lastNum = $num;
        }

        return $firstNum . $lastNum;
    }

    return "";
}

foreach ($lines as $line) {
    // Part 1
    preg_match_all('/\d/', $line, $matches);
    $digits = $matches[0];

    // Part 2
    if (!empty($digits)) {
        $firstDigit = $digits[0];
        $lastDigit = end($digits);
        $r1 += intval($firstDigit . $lastDigit);
    }

    $num_val = get_num($line);
    $r2 += (int) $num_val;
}

echo "p1: {$r1} p2: {$r2}";

File: 2023/Day 1/perl/day1.pl
#!/usr/bin/perl

use strict;
use warnings;

my %num_words_to_digits = (
    'zero'  => 0,
    'one'   => 1,
    'two'   => 2,
    'three' => 3,
    'four'  => 4,
    'five'  => 5,
    'six'   => 6,
    'seven' => 7,
    'eight' => 8,
    'nine'  => 9
);

sub main {
    open my $fh, '<', 'input.txt'
      or die "Could not open file: $!";
    my ( $r1_total, $r2_total ) = ( 0, 0 );

    while ( my $line = < $fh > ) {
        chomp $line;

        # Part 1
        my @digits = $line = ~/(\d)/g;
        $r1_total += sum_first_last(@digits) if @digits;

        # Part 2
        $line = replace_number_words( $line, \%num_words_to_digits );
        @digits = $line = ~/(\d)/g;
        $r2_total += sum_first_last(@digits) if @digits;

    }
    print "p1: $r1_total, p2: $r2_total\n";
    close $fh;
}

sub sum_first_last {
    my @nums = @_;
    return $nums[0] . ( $nums[-1] // $nums[0] );
}

sub replace_number_words {
    my $line = shift;

    my $pattern = join( "|", keys %num_words_to_digits, values %num_words_to_digits );
    my @matches = $line = ~/($pattern)/g;

    if (@matches) {
        my $firstNum =
          defined $num_words_to_digits{ $matches[0] }
          ? $num_words_to_digits{ $matches[0] }
          : $matches[0];
        my $lastNum =
          defined $num_words_to_digits{ $matches[-1] }
          ? $num_words_to_digits{ $matches[-1] }
          : $matches[-1];

        return $firstNum . $lastNum;
    }

    return "";
}

main();
